accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,113 @@
+
+### Which components own the mapping between different representations of feature switches
+There are 3 different representations to each feature switch
+* The MSBuild property - for example `InvariantGlobalization` - this is used by developers to turn the feature on/off.
+* The runtime property in `.runtimeconfig.json` - for example `System.Globalization.Invariant` - this is used at runtime to alter the behavior of the code (via the IL property below)
+* The IL property which is used in the code - for example `System.Globalization.GlobalizationMode.Invariant, System.Private.CoreLib` - this is used by the linker to fix the value of the property and propagate that to remove unused code.
+
+The mapping between MSBuild property and runtime property pretty much has to be part of the SDK (in the MSBuild somewhere) as the SDK already generates the `.runtimeconfig.json` and this needs to be part of it. The second mapping to the IL property can take several options:
+* Also int he SDK - becomes an input to the ILLink task
+* Encoded as attributes in the BCL - we could add a new attribute to the IL properties to mark which runtime options they map to - linker can create the mapping from these attributes. In this case the ILLink task would take the runtime properties as input
+* Inferred by the linker - in the extreme linker might be able to track the calls to `AppContext.TryGetSwitch` and perform enough data flow analysis to determine the value of the IL property from the runtime property value. In this case the ILLink task would also take runtime properties as input
+
+### Testing challenges
+Having feature switches which are exposed as public knobs means that the number of valid and supported configurations of libraries grows in a significant way. Testing each feature switch in isolation is not that costly, but making sure there are no unintended interactions might be problematic.
+
+### Stability
+Using `AppContext.TryGetSwitch` as the underlying runtime store of these switches is consistent with the intent of that API (and the API has been present in the .NET libraries for a long time). Unfortunately there are no guarantees that the value will never change, in fact there's an API `AppContext.SetSwitch` which is designed to change the value at runtime. Fixing the values of these switches in the linker would make usage of the `SetSwitch` API ineffective and rather confusing.
+
+*Note that the existing compatibility switches in `CoreLib` already have this problem and the current design seems to be to ignore it. That is the value is read from `AppContext` the first time it's needed and cached.*
+
+
+## References
+
+Very similar functionality has been proposed before in [dotnet/designs#42](https://github.com/dotnet/designs/pull/42). The main differences proposed here:
+* Feature switches apply at runtime as well as in the linker. The previous proposal was only for linker.
```

Comment:
The link to AppContext so that we can emulate this at runtime was in that proposal too: https://github.com/dotnet/designs/pull/42/files#diff-416341d2703781003936b3f17e39fd0eR89 😉

Reply:
Sorry - totally missed that - fixed.